### DISCUSSION TITLE
Read-only buffer support

### DISCUSSION THREAD

**Discussion by heaths** [helix-editor/helix#11586] [Ideas]
In #7128 the idea was brought up to have a read-only mode similar to vim. @the-mikedavis also mentioned Kakoune as an example that refuses to enter Insert mode, compared to vim that simply warns you you're trying to edit a read-only buffer. I can't find an existing discussion for this, so I wanted to start one here because I like to use vim often as a better pager e.g., `curl https://domain.tld/json | jq | vim -R -`.

Personally, I like vim's behavior that is warns I'm editing a read-only buffer and then let whether there's even a file path attached to the buffer or the file is protected decide whether I can save it. This lends itself well to other use e.g., certain user-writable buffers like help content could automatically be marked read-only but we could forcibly overwrite them if we have suitable permissions. I.e., don't make helix the gatekeeper for whether we can do something - only let us know it may not work.

**Comment by Deebster** [comment]
+1

My ideal for this feature would be that it allows you to edit (with a visible warning), but doesn't complain about the unsaved buffer. This would allow both saving (when desired) and minimal friction when using as a pager.

edit: So I guess "read-only buffer" is a misnomer. Ephemeral buffer? Volatile buffer? Transient buffer?

**Reply by eikenb** [reply]
> edit: So I guess "read-only buffer" is a misnomer. Ephemeral buffer? Volatile buffer? Transient buffer?

I'd say this is more like a buffer for a read-only file. Could probably reuse the read-only-indicator for this as well.

**Comment by Guillawme** [comment]
I would also happily use helix in read-only mode as a pager. Its navigation commands would be great for this use-case.

**Comment by tastapod** [comment]
I think vi and vim get the terminology wrong here. Using `vim -R` or `view` opens the buffer in a _write-protected mode_ that they call readonly.

You can do all the editing you like, but when you go to close the buffer, you have to use either `w!` or `q!`. In other words, you have to deliberately force-write or abandon any changes you made (or write them to a new buffer).

I feel like this would be trivial to implement in helix, given it already recognizes write-protected files when you open them. It would just be a per-buffer toggle.

- `:set readonly true` would match the `[readonly]` marker when you open a write-protected file.
- `:set write-protect true` would align better with other settings (there are several `write` settings, and nothing about `read` or `readonly`).
- `:set save-protect true` would match other `save` settings (e.g. auto-saving options).

That said, settings tend to be session-wide rather than buffer-specific, so maybe:

- `:buffer-protect` as a toggle to match e.g. `buffer-close`

Regardless, I would love to see this as a drop-in replacement for `view`.
